### PR TITLE
Dev multi

### PR DESCRIPTION
**Pull request by kirillProkofiev**
start cleaning and preparing the code for the OTE.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -3,216 +3,370 @@
...
+# parse arguments
+parser.add_argument('--GPU', type=int, default=1, help='specify which gpu to use')
+parser.add_argument('--print-freq', '-p', default=5, type=int, help='print frequency (default: 20)')
+parser.add_argument('--save_checkpoint', type=bool, default=True, help='whether or not to save your model')
+parser.add_argument('--config', type=str, default='config.py', required=True,
+                        help='Configuration file')
+
+# global variables and argument parsing
+args = parser.parse_args()
+path_to_config = os.path.join(current_dir, args.config)
```
Comment (sovrasov) [inline train.py:28]: The same thing as in other scripts: config can be located anywhere.

File: train.py [hunk 2]
```diff
@@ -3,216 +3,370 @@
...
+config = read_py_config(path_to_config)
+experiment_snapshot = config['checkpoint']['snapshot_name']
+experiment_path = config['checkpoint']['experiment_path']
+WRITER = SummaryWriter(experiment_path)
+STEP, VAL_STEP = 0, 0
+BEST_ACCURACY, CURRENT_ACCURACY, BEST_AUC, BEST_EER, BEST_ACER = 0, 0, 0, float('inf'), float('inf')
```
Comment (sovrasov) [inline train.py:34]: Try to avoid global variables.

File: train.py [hunk 3]
```diff
@@ -3,216 +3,370 @@
...
+
+        # filter output for real images and compute third loss
+        mask = target[:,0] == 0
+        filtered_output = output[3][mask] 
+        filtered_target = target[:,3:][mask].type(torch.float32)
+        real_atr_loss = BCE(filtered_output, filtered_target)
+        
+    # taking derivitives
+    optimizer.zero_grad()
+    spoof_loss.backward(retain_graph=True)
```
Comment (sovrasov) [inline train.py:316]: You actually need to call `backward()` only once on the resulting sum of the losses.

File: train.py [hunk 4]
```diff
@@ -3,216 +3,370 @@
...
+    optimizer.zero_grad()
+    lightning_loss.backward(retain_graph=True)
+
+    optimizer.zero_grad()
+    real_atr_loss.backward(retain_graph=True)
+    # combine losses
+    loss = C*spoof_loss + Cs*spoof_type_loss + Ci*lightning_loss + Cf*real_atr_loss
+    return loss
+
+def init_experiment(config, path_to_config):
```
Comment (sovrasov) [inline train.py:330]: Misleading name of the function

File: train.py [hunk 5]
```diff
@@ -3,216 +3,370 @@
...
+            save_checkpoint(checkpoint, f'{experiment_path}/{experiment_snapshot}')
+            
+    # evaulate in the end of training    
+    if config['evaulation']:
+        eval_model(model, config, val_transform, map_location = args.GPU, eval_func = evaulate, file_name='LCC_FASD.txt', flag=None)
+        eval_model(model, config, val_transform, map_location = args.GPU, eval_func = evaulate, file_name='Celeba_test.txt', flag=True)
 
 def train(train_loader, model, criterion, optimizer, epoch):
-    global STEP, args
+    global STEP, args, config
```
Comment (sovrasov) [inline train.py:147]: To get rid of unclear manipulations with global variables I suggest to create a class for training that will contain training loops, eval methods and `mixup_criterion` function as a method.

File: utils.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+        output = output*s
+    if type(output) == tuple:
+        output = output[0].data
+    accuracy = (output.argmax(dim=1) == target).float().mean().item()
+    return accuracy*100
+
+def mixup_target(input, target, config, cuda, num_classes=2):
+    # compute mix-up augmentation
+    input, target_a, target_b, lam = mixup_data(input, target, config['aug']['alpha'], config['aug']['beta'], cuda)
+    input, target_a, target_b = map(Variable, (input, target_a, target_b))
```
Comment (sovrasov) [inline utils.py:82]: `Variable` is a class to emphasize tensors that require grad an ancient versions of pytorch. Now it's completely not needed and exists only for backward compatibility.

File: utils.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+                                                shuffle=True, pin_memory=config['data']['pin_memory'],
+                                                num_workers=config['data']['data_loader_workers'])
+    return train_loader, val_loader
+
+def build_model(config, args, strict=True):
+    ''' build model and change layers depends on loss type'''
+
+    if config['model']['model_type'] == 'Mobilenet2':
+        raise NotImplementedError
+        # model = mobilenetv2(prob_dropout=config['dropout']['prob_dropout'])
```
Comment (sovrasov) [inline utils.py:194]: Dead code

**Review by sovrasov — CHANGES_REQUESTED** [review]

File: configs/config.py [hunk 8]
```diff
@@ -1,41 +1,71 @@
 dataset = 'celeba-spoof'
 
-datasets = dict(LCCFASD_root='/home/prokofiev/pytorch/LCC_FASDnew', Celeba_root='/home/prokofiev/pytorch/antispoofing/CelebA_Spoof')
+multi_task_learning = True
+
+evaulation = True
+
+datasets = dict(LCCFASD_root='/home/prokofiev/pytorch/LCC_FASDnew',
```
Comment (sovrasov) [inline configs/config.py:7]: Get rid of hardcoded absolute path within the project

Comment (kirillProkofiev) [reply configs/config.py:7]: This is config file. How can I specify path to my folder without explicitly write it?

Comment (sovrasov) [reply configs/config.py:7]: This is actually a sample config provided for users. Paths like `/home/vasyan/my_first_pytorch_project` don't provide any useful information for them.

File: demo/demo.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        self.model = model
+        if config.data_parallel.use_parallel.:
+            self.model = nn.DataParallel(self.model, **config.data_parallel.parallel_params.)
+        utils.load_checkpoint(checkpoint_path, self.model, map_location=device)
+        self.config = config
+
+    def preprocessing(self, images):
+        ''' making image preprocessing for pytorch pipeline '''
+        mean = np.array(self.config.img_norm_cfg.mean.).reshape(3,1,1)
+        std = np.array(self.config.img_norm_cfg.std.).reshape(3,1,1)
```
Comment (sovrasov) [inline demo/demo.py:104]: Why do you need an extra dot here? (`mean.`)

Comment (kirillProkofiev) [reply demo/demo.py:104]: bug. I will fix this.

File: README.md [hunk 10]
```diff
@@ -26,24 +25,27 @@ bash init_venv.sh
...
 ```
 ### Data Preparation
-For training or evaluating on CelebA Spoof dataset you need to download dataset (you can do it from their official repository) and then run following script being located in root folder of the project:
+For training or evaluating on CelebA Spoof dataset you need to download dataset (you can do it from their [official repository](https://github.com/Davidzhangyuanhan/CelebA-Spoof)) and then run following script being located in root folder of the project:
 ```bash
 cd /data_preparation/
 python3 prepare_celeba_json.py
 ```
-To train or evaluate on LCC FASD dataset you need to download it (link is available in their paper on [arxive]()) and run following script:
+To train or evaluate on LCC FASD dataset you need to download it (link is available in their paper on [arxive](https://csit.am/2019/proceedings/PRIP/PRIP3.pdf)) and run following script:
```
Comment (sovrasov) [inline README.md:33]: arxiv

File: README.md [hunk 11]
```diff
@@ -26,24 +25,27 @@ bash init_venv.sh
...
 ### Configuration file
-The script for training and inference uses a configuration file. This is default [configuration file](). You need to specify paths to datasets. Training pipeline supports following methods, which you can switch on and tune hyperparams while training:
+The script for training and inference uses a configuration file. This is default [configuration file](./configs/config.py). You need to specify paths to datasets. Training pipeline supports following methods, which you can switch on and tune hyperparams while training:
 * RSC - representation self challenging, applied before global average pooling. p, b - quantile and probability applying it on image in current batch
 * aug - advanced augmentation, appropriate value for type is 'cutmix' or 'mixup. lambda = BetaDistribution(alpha, beta), cutmix_prob - probability of applying cutmix on image.
-...
-* ...
+* loss - there are available two possible losses: 'amsoftmax' with 'cos','arcos','cross_enropy' margins and 'soft_triple' with different number of inner classes. For more details about this soft loss see in [paper](https://arxiv.org/pdf/1909.05235.pdf)
```
Comment (sovrasov) [inline README.md:44]: soft triple?

File: README.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+* data_parallel - you can train your network on several GPU
+* multi_task_learning - specify whether or not to train with multitask loss
+* conv_cd - this is option to switch on central difference convolutions instead of vanilla one changing value of theta from 0
+
+## Training
+To start training create config file based on the default one and run 'train.py':
+```bash
+sudo python3 train.py --config <path to config>;
+```
+For additional parameters you can refer to help adding '--help'. For example, you can specify on which GPU you want to train your model. If for some reasons you want to train on CPU, specify `--device` to 'cpu'. default device is cuda 0. As you may notice, training pipeline supports parallel training and specified GPU in arguments must be the same as in configuration file for output GPU.
```
Comment (sovrasov) [inline README.md:59]: What if the devices specified via config and command line are different? I think there should be a rule that defines how to override the config in such case or all the device selection parameters should be located in config only.

File: README.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+```bash
+bash init_venv.sh
+```
+
+2. Activate virtual environment:
+```bash
+. venv/bin/activate
+```
+### Data Preparation
+For training or evaluating on CelebA Spoof dataset you need to download dataset (you can do it from their [official repository](https://github.com/Davidzhangyuanhan/CelebA-Spoof)) and then run following script being located in root folder of the project:
```
Comment (sovrasov) [inline README.md:30]: from the, run the following script, in the root folder

File: README.md [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+```bash
+cd /data_preparation/
+python prepare_celeba_json.py
+```
+To train or evaluate on LCC FASD dataset you need to download it (link is available in their paper on [arxiv](https://csit.am/2019/proceedings/PRIP/PRIP3.pdf)). Then you need to get the OpenVINO™ face detector model (choose the one on the goole drive), activate OpenVINO™ enviroment and run following script:
```
Comment (sovrasov) [inline README.md:35]: link is available in the [original paper]

File: README.md [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+| model name | dataset | AUC | EER% | APCER% | BPCER% | ACER% | MParam | GFlops | Link to snapshot |
+| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| MN3_large |CelebA-Spoof| 0.998 | 2.26 | 0.69 | 6.92 | 3.8 | 0.13 | 2.93 | [snapshot](https://drive.google.com/drive/u/0/folders/1A6wa3AlrdjyNPkXT81knIzXxR7SAYm1q) |
```
Comment (sovrasov) [inline README.md:7]: Access to the folder is restricted

File: README.md (hunk 14 above)
Comment (sovrasov) [inline README.md:35]: As far as I remember we decided not to duplicate IR with FD on google drive and instead insert a reference to the [model downloader](https://docs.openvinotoolkit.org/latest/omz_tools_downloader_README.html) tool.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+This script will cut faces tighter than it is in original dataset and get rid of some garbage crops. For running this script you need to activate OpenVINO™ environment. Refer to official documentation.
+
+You can use LCC FASD without doing this at all, but it seems to enhance performance, so I recommend doing this.
+Note that the new folder will be created and named as `<old name>cropped`. So to train or test model with cropped data, please, set path to that new folder, which will be located in the same directory as launched script.
+
+To train or evaluate on CASIA CEFA you just need to download it. Reader for this dataset supports not only RGB modality, but depth and IR too. Nevertheless, it's not the purpose of this project.
+
+If you want to use your own data, the next steps should be done:
+1) prepare reader for your dataset.
+2) Import reader object to datasets/database.py file. Substitute `do_nothing` with your object in `external_reader=do_nothing` (35 line).
```
Comment (sovrasov) [inline README.md:48]: Inconsistent case of the first letter

File: README.md [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+3) in config, write any kwargs for train, val, test protocol. If you do not have test data, you can just add the same parameters as in val.
+Example: `external = dict(train=dict(data_root='...', mode='train', whatever=...), val=dict(data_root='...', mode='val', whatever=...), test=dict(...))`
+
+Now you are ready to launch training process!
+
+### Configuration file
+The script for training and inference uses a configuration file. This is default [configuration file](./configs/config.py). You need to specify paths to datasets. Training pipeline supports following methods, which you can switch on and tune hyperparams while training:
```
Comment (sovrasov) [inline README.md:55]: This  is the default one

File: README.md [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+python convert_model.py --config <path to config>; --model_path <path to where save the model>;
+```
+By default, the output model path is 'MobileNetv3.onnx'
+Now you obtain '.onnx' format. To obtain OpenVINO™ IR model you should refer to [official documentation](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_Deep_Learning_Model_Optimizer_DevGuide.html) for next steps. You do not need to add --mean, --scale, --inverse_input channels options.
+To check that there are no mistakes with the conversion you can launch `conversion_checker.py` by writing following command:
+```bash
+python conversion_checker.py --config <path to config>; --spf_model_torch <path to torch model> --spf_model_openvino <path to openvino model>;
+```
+You will see mean difference (L1 metric distance) on the first and second predicted class. If it's 10e-6 or less than it's all good.
+
```
Comment (sovrasov) [inline README.md:105]: Please add also a step of conversion to IR with the model optimizer.

File: README.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+Use the `mo_onnx.py` script from the <INSTALL_DIR>/deployment_tools/model_optimizer directory to run the Model Optimizer:
+```bash
+python mo_onnx.py --input_model <INPUT_MODEL.onnx> --mean_values [151.2405,119.595,107.8395] --scale_values [63.0105,56.457,55.0035] --reverse_input_channels
+```
+Note, that parameters of the mean and scale values should be in the [0,255] range (byte format)
+To check that there are no mistakes with the conversion you can launch `conversion_checker.py` by writing the following command:
+```bash
+python conversion_checker.py --config <path to config>; --spf_model_torch <path to torch model> --spf_model_openvino <path to OpenVINO model>;
+```
+You will see the mean difference (L1 metric distance) on the first and second predicted class. If it's 10e-6 or less than it's all good.
```
Comment (sovrasov) [inline README.md:113]: then?

**Review by sovrasov — APPROVED** [review]
